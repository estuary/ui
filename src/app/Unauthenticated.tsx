import Auth from 'pages/Auth';
import Login from 'pages/Login';
import { Route, Routes } from 'react-router';
import { unauthenticatedRoutes } from './routes';

const Unauthenticated = () => {
    console.log('Unauthenticated');

    return (
        <Routes>
            <Route path={unauthenticatedRoutes.auth.path} element={<Auth />} />
            <Route
                path={unauthenticatedRoutes.magicLink.path}
                element={<Auth />}
            />
            <Route path="*" element={<Login />} />
        </Routes>
    );
};

export default Unauthenticated;
