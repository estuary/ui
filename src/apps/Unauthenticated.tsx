import Login from 'pages/Login';

const Unauthenticated = () => {
    return <Login />;
};

export default Unauthenticated;
