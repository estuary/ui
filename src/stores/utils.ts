import { isEmpty, map } from 'lodash';

export const checkForErrors = (obj: any) =>
    obj?.errors && obj.errors.length > 0;

export const populateErrors = (configDictionary: any) => {
    let configErrors: any[] = [];

    // TODO (errors) When there are not configs do we need to populate this object with something?
    if (Object.keys(configDictionary).length > 0) {
        map(configDictionary, (config) => {
            if (checkForErrors(config)) {
                configErrors = configErrors.concat(config.errors);
            }
        });
    }

    return {
        configErrors,
        hasErrors: !isEmpty(configErrors),
    };
};
