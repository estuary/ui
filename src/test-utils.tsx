// https://testing-library.com/docs/react-testing-library/setup#custom-render

import { render, RenderOptions } from '@testing-library/react';
import { ReactElement, type FC } from 'react';
import { act } from 'react-dom/test-utils';
import { localStorageKey } from './auth';
import AppProviders from './context';

const AllTheProviders: FC = ({ children }) => {
    return <AppProviders>{children}</AppProviders>;
};

const customRender = (
    ui: ReactElement,
    options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

const loginAsUser = (userName: string = 'fakeUserName') => {
    act(() => {
        window.localStorage.setItem(localStorageKey, userName);
    });
};

export * from '@testing-library/react';
export { customRender as render, loginAsUser };
