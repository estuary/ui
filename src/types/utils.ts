export type WithRequiredProperty<T, K extends keyof T> = T & {
    [P in K]-?: T[P];
};

export type WithRequiredNonNullProperty<T, K extends keyof T> = T & {
    [P in K]-?: Exclude<T[P], null>;
};
