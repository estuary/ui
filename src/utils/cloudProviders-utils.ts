// Tried our best to pick flags that make sense for users. Did not use state, provincial, etc. flags
//  to keep this simpler. The flags were chosen by finding the country on wikipedia's list:
//      https://en.wikipedia.org/wiki/Regional_indicator_symbol

const defaultResponse = {
    location: `unknown`,
    flag: `?`,
};

// from https://docs.aws.amazon.com/global-infrastructure/latest/regions/aws-regions.html
const aws = {
    'us-east-1': {
        location: `US East (N. Virginia), United States of America`,
        flag: `ğŸ‡ºğŸ‡¸`,
    },
    'us-east-2': {
        location: `US East (Ohio), United States of America`,
        flag: `ğŸ‡ºğŸ‡¸`,
    },
    'us-west-1': {
        location: `US West (N. California), United States of America`,
        flag: `ğŸ‡ºğŸ‡¸`,
    },
    'us-west-2': {
        location: `US West (Oregon), United States of America`,
        flag: `ğŸ‡ºğŸ‡¸`,
    },
    'af-south-1': {
        location: `Africa (Cape Town), South Africa`,
        flag: `ğŸ‡¿ğŸ‡¦`,
    },
    'ap-east-1': {
        location: `Asia Pacific (Hong Kong), Hong Kong`,
        flag: `ğŸ‡­ğŸ‡°`,
    },
    'ap-south-2': {
        location: `Asia Pacific (Hyderabad), India`,
        flag: `ğŸ‡®ğŸ‡³`,
    },
    'ap-southeast-3': {
        location: `Asia Pacific (Jakarta), Indonesia`,
        flag: `ğŸ‡®ğŸ‡©`,
    },
    'ap-southeast-5': {
        location: `Asia Pacific (Malaysia), Malaysia`,
        flag: `ğŸ‡²ğŸ‡¾`,
    },
    'ap-southeast-4': {
        location: `Asia Pacific (Melbourne), Australia`,
        flag: `ğŸ‡¦ğŸ‡º`,
    },
    'ap-south-1': { location: `Asia Pacific (Mumbai), India`, flag: `ğŸ‡®ğŸ‡³` },
    'ap-northeast-3': { location: `Asia Pacific (Osaka), Japan`, flag: `ğŸ‡¯ğŸ‡µ` },
    'ap-northeast-2': {
        location: `Asia Pacific (Seoul), South Korea`,
        flag: `ğŸ‡°ğŸ‡·`,
    },
    'ap-southeast-1': {
        location: `Asia Pacific (Singapore), Singapore`,
        flag: `ğŸ‡¸ğŸ‡¬`,
    },
    'ap-southeast-2': {
        location: `Asia Pacific (Sydney), Australia`,
        flag: `ğŸ‡¦ğŸ‡º`,
    },
    'ap-southeast-7': {
        location: `Asia Pacific (Thailand), Thailand`,
        flag: `ğŸ‡¹ğŸ‡­`,
    },
    'ap-northeast-1': { location: `Asia Pacific (Tokyo), Japan`, flag: `ğŸ‡¯ğŸ‡µ` },
    'ca-central-1': { location: `Canada (Central), Canada`, flag: `ğŸ‡¨ğŸ‡¦` },
    'ca-west-1': { location: `Canada West (Calgary), Canada`, flag: `ğŸ‡¨ğŸ‡¦` },
    'eu-central-1': { location: `Europe (Frankfurt), Germany`, flag: `ğŸ‡©ğŸ‡ª` },
    'eu-west-1': { location: `Europe (Ireland), Ireland`, flag: `ğŸ‡®ğŸ‡ª` },
    'eu-west-2': { location: `Europe (London), United Kingdom`, flag: `ğŸ‡¬ğŸ‡§` },
    'eu-south-1': { location: `Europe (Milan), Italy`, flag: `ğŸ‡®ğŸ‡¹` },
    'eu-west-3': { location: `Europe (Paris), France`, flag: `ğŸ‡«ğŸ‡·` },
    'eu-south-2': { location: `Europe (Spain), Spain`, flag: `ğŸ‡ªğŸ‡¸` },
    'eu-north-1': { location: `Europe (Stockholm), Sweden`, flag: `ğŸ‡¸ğŸ‡ª` },
    'eu-central-2': { location: `Europe (Zurich), Switzerland`, flag: `ğŸ‡¨ğŸ‡­` },
    'il-central-1': { location: `Israel (Tel Aviv), Israel`, flag: `ğŸ‡®ğŸ‡±` },
    'mx-central-1': { location: `Mexico (Central), Mexico`, flag: `ğŸ‡²ğŸ‡½` },
    'me-south-1': { location: `Middle East (Bahrain), Bahrain`, flag: `ğŸ‡§ğŸ‡­` },
    'me-central-1': {
        location: `Middle East (UAE), United Arab Emirates`,
        flag: `ğŸ‡¦ğŸ‡ª`,
    },
    'sa-east-1': { location: `South America (SÃ£o Paulo), Brazil`, flag: `ğŸ‡§ğŸ‡·` },
};

// from https://gist.github.com/ausfestivus/04e55c7d80229069bf3bc75870630ec8
const azure = {
    southafricanorth: { location: `(Africa) South Africa North`, flag: `ğŸ‡¿ğŸ‡¦` },
    southafricawest: { location: `(Africa) South Africa West`, flag: `ğŸ‡¿ğŸ‡¦` },
    australiacentral: {
        location: `(Asia Pacific) Australia Central`,
        flag: `ğŸ‡¦ğŸ‡º`,
    },
    australiacentral2: {
        location: `(Asia Pacific) Australia Central 2`,
        flag: `ğŸ‡¦ğŸ‡º`,
    },
    australiaeast: { location: `(Asia Pacific) Australia East`, flag: `ğŸ‡¦ğŸ‡º` },
    australiasoutheast: {
        location: `(Asia Pacific) Australia Southeast`,
        flag: `ğŸ‡¦ğŸ‡º`,
    },
    centralindia: { location: `(Asia Pacific) Central India`, flag: `ğŸ‡®ğŸ‡³` },
    // 'eastasia': {location:`(Asia Pacific) East Asia`, flag:`?`},
    japaneast: { location: `(Asia Pacific) Japan East`, flag: `ğŸ‡¯ğŸ‡µ` },
    japanwest: { location: `(Asia Pacific) Japan West`, flag: `ğŸ‡¯ğŸ‡µ` },
    jioindiacentral: {
        location: `(Asia Pacific) Jio India Central`,
        flag: `ğŸ‡®ğŸ‡³`,
    },
    jioindiawest: { location: `(Asia Pacific) Jio India West`, flag: `ğŸ‡®ğŸ‡³` },
    koreacentral: { location: `(Asia Pacific) Korea Central`, flag: `ğŸ‡®ğŸ‡³` },
    koreasouth: { location: `(Asia Pacific) Korea South`, flag: `ğŸ‡°ğŸ‡·` },
    southindia: { location: `(Asia Pacific) South India`, flag: `ğŸ‡®ğŸ‡³` },
    // 'southeastasia': {location:`(Asia Pacific) Southeast Asia`, flag:`?`},
    westindia: { location: `(Asia Pacific) West India`, flag: `ğŸ‡®ğŸ‡³` },
    canadacentral: { location: `(Canada) Canada Central`, flag: `ğŸ‡¨ğŸ‡¦` },
    canadaeast: { location: `(Canada) Canada East`, flag: `ğŸ‡¨ğŸ‡¦` },
    francecentral: { location: `(Europe) France Central`, flag: `ğŸ‡«ğŸ‡·` },
    francesouth: { location: `(Europe) France South`, flag: `ğŸ‡«ğŸ‡·` },
    germanynorth: { location: `(Europe) Germany North`, flag: `ğŸ‡©ğŸ‡ª` },
    germanywestcentral: {
        location: `(Europe) Germany West Central`,
        flag: `ğŸ‡©ğŸ‡ª`,
    },
    italynorth: { location: `(Europe) Italy North`, flag: `ğŸ‡®ğŸ‡¹` },
    northeurope: { location: `(Europe) North Europe`, flag: `ğŸ‡ªğŸ‡º` },
    norwayeast: { location: `(Europe) Norway East`, flag: `ğŸ‡³ğŸ‡´` },
    norwaywest: { location: `(Europe) Norway West`, flag: `ğŸ‡³ğŸ‡´` },
    polandcentral: { location: `(Europe) Poland Central`, flag: `ğŸ‡µğŸ‡±` },
    spaincentral: { location: `(Europe) Spain Central`, flag: `ğŸ‡ªğŸ‡¸` },
    swedencentral: { location: `(Europe) Sweden Central`, flag: `ğŸ‡¸ğŸ‡ª` },
    switzerlandnorth: { location: `(Europe) Switzerland North`, flag: `ğŸ‡¨ğŸ‡­` },
    switzerlandwest: { location: `(Europe) Switzerland West`, flag: `ğŸ‡¨ğŸ‡­` },
    uksouth: { location: `(Europe) UK South`, flag: `ğŸ‡¬ğŸ‡§` },
    ukwest: { location: `(Europe) UK West`, flag: `ğŸ‡¬ğŸ‡§` },
    westeurope: { location: `(Europe) West Europe`, flag: `ğŸ‡ªğŸ‡º` },
    mexicocentral: { location: `(Mexico) Mexico Central`, flag: `ğŸ‡²ğŸ‡½` },
    israelcentral: { location: `(Middle East) Israel Central`, flag: `ğŸ‡®ğŸ‡±` },
    qatarcentral: { location: `(Middle East) Qatar Central`, flag: `ğŸ‡¶ğŸ‡¦` },
    uaecentral: { location: `(Middle East) UAE Central`, flag: `ğŸ‡¦ğŸ‡ª` },
    uaenorth: { location: `(Middle East) UAE North`, flag: `ğŸ‡¦ğŸ‡ª` },
    brazilsouth: { location: `(South America) Brazil South`, flag: `ğŸ‡§ğŸ‡·` },
    brazilsoutheast: {
        location: `(South America) Brazil Southeast`,
        flag: `ğŸ‡§ğŸ‡·`,
    },
    brazilus: { location: `(South America) Brazil US`, flag: `ğŸ‡§ğŸ‡·` },
    centralus: { location: `(US) Central US`, flag: `ğŸ‡ºğŸ‡¸` },
    centraluseuap: { location: `(US) Central US EUAP`, flag: `ğŸ‡ºğŸ‡¸` },
    eastus: { location: `(US) East US`, flag: `ğŸ‡ºğŸ‡¸` },
    eastus2: { location: `(US) East US 2`, flag: `ğŸ‡ºğŸ‡¸` },
    eastus2euap: { location: `(US) East US 2 EUAP`, flag: `ğŸ‡ºğŸ‡¸` },
    eastusstg: { location: `(US) East US STG`, flag: `ğŸ‡ºğŸ‡¸` },
    northcentralus: { location: `(US) North Central US`, flag: `ğŸ‡ºğŸ‡¸` },
    southcentralus: { location: `(US) South Central US`, flag: `ğŸ‡ºğŸ‡¸` },
    southcentralusstg: { location: `(US) South Central US STG`, flag: `ğŸ‡ºğŸ‡¸` },
    westcentralus: { location: `(US) West Central US`, flag: `ğŸ‡ºğŸ‡¸` },
    westus: { location: `(US) West US`, flag: `ğŸ‡ºğŸ‡¸` },
    westus2: { location: `(US) West US 2`, flag: `ğŸ‡ºğŸ‡¸` },
    westus3: { location: `(US) West US 3`, flag: `ğŸ‡ºğŸ‡¸` },
};

// from https://cloud.google.com/compute/docs/regions-zones#available
const gcp = {
    'africa-south1': { location: `Johannesburg, South Africa`, flag: `ğŸ‡¿ğŸ‡¦` },
    'asia-east1': { location: `Changhua County, Taiwan, APAC`, flag: `ğŸ‡¹ğŸ‡¼` },
    'asia-east2': { location: `Hong Kong, APAC`, flag: `ğŸ‡­ğŸ‡°` },
    'asia-northeast1': { location: `Tokyo, Japan, APAC`, flag: `ğŸ‡¯ğŸ‡µ` },
    'asia-northeast2': { location: `Osaka, Japan, APAC`, flag: `ğŸ‡¯ğŸ‡µ` },
    'asia-northeast3': { location: `Seoul, South Korea, APAC`, flag: `ğŸ‡°ğŸ‡·` },
    'asia-south1': { location: `Mumbai, India, APAC`, flag: `ğŸ‡®ğŸ‡³` },
    'asia-south2': { location: `Delhi, India, APAC`, flag: `ğŸ‡®ğŸ‡³` },
    'asia-southeast1': {
        location: `Jurong West, Singapore, APAC`,
        flag: `ğŸ‡¸ğŸ‡¬`,
    },
    'asia-southeast2': { location: `Jakarta, Indonesia, APAC`, flag: `ğŸ‡®ğŸ‡©` },
    'australia-southeast1': { location: `Sydney, Australia, APAC`, flag: `ğŸ‡¦ğŸ‡º` },
    'australia-southeast2': {
        location: `Melbourne, Australia, APAC`,
        flag: `ğŸ‡¦ğŸ‡º`,
    },
    'europe-central2': { location: `Warsaw, Poland, Europe`, flag: `ğŸ‡µğŸ‡±` },
    'europe-north1': { location: `Hamina, Finland, Europe`, flag: `ğŸ‡«ğŸ‡®` },
    'europe-north2': { location: `Stockholm, Sweden, Europe`, flag: `ğŸ‡¸ğŸ‡ª` },
    'europe-southwest1': { location: `Madrid, Spain, Europe`, flag: `ğŸ‡ªğŸ‡¸` },
    'europe-west1': { location: `St. Ghislain, Belgium, Europe`, flag: `ğŸ‡§ğŸ‡ª` },
    'europe-west10': { location: `Berlin, Germany, Europe`, flag: `ğŸ‡©ğŸ‡ª` },
    'europe-west12': { location: `Turin, Italy, Europe`, flag: `ğŸ‡®ğŸ‡¹` },
    'europe-west2': { location: `London, England, Europe`, flag: `ğŸ‡¬ğŸ‡§` },
    'europe-west3': { location: `Frankfurt, Germany, Europe`, flag: `ğŸ‡©ğŸ‡ª` },
    'europe-west4': { location: `Eemshaven, Netherlands, Europe`, flag: `ğŸ‡³ğŸ‡±` },
    'europe-west6': { location: `Zurich, Switzerland, Europe`, flag: `ğŸ‡¨ğŸ‡­` },
    'europe-west8': { location: `Milan, Italy, Europe`, flag: `ğŸ‡®ğŸ‡¹` },
    'europe-west9': { location: `Paris, France, Europe`, flag: `ğŸ‡«ğŸ‡·` },
    'me-central1': { location: `Doha, Qatar, Middle East`, flag: `ğŸ‡¶ğŸ‡¦` },
    'me-central2': {
        location: `Dammam, Saudi Arabia, Middle East`,
        flag: `ğŸ‡¸ğŸ‡¦`,
    },
    'me-west1': { location: `Tel Aviv, Israel, Middle East`, flag: `ğŸ‡®ğŸ‡±` },
    'northamerica-northeast1': {
        location: `MontrÃ©al, QuÃ©bec, North America`,
        flag: `ğŸ‡¨ğŸ‡¦`,
    },
    'northamerica-northeast2': {
        location: `Toronto, Ontario, North America`,
        flag: `ğŸ‡¨ğŸ‡¦`,
    },
    'northamerica-south1': {
        location: `Queretaro, Mexico, North America`,
        flag: `ğŸ‡²ğŸ‡½`,
    },
    'southamerica-east1': {
        location: `Osasco, SÃ£o Paulo, Brazil, South America`,
        flag: `ğŸ‡§ğŸ‡·`,
    },
    'southamerica-west1': {
        location: `Santiago, Chile, South America`,
        flag: `ğŸ‡¨ğŸ‡±`,
    },
    'us-central1': {
        location: `Council Bluffs, Iowa, North America`,
        flag: `ğŸ‡ºğŸ‡¸`,
    },
    'us-east1': {
        location: `Moncks Corner, South Carolina, North America`,
        flag: `ğŸ‡ºğŸ‡¸`,
    },
    'us-east4': { location: `Ashburn, Virginia, North America`, flag: `ğŸ‡ºğŸ‡¸` },
    'us-east5': { location: `Columbus, Ohio, North America`, flag: `ğŸ‡ºğŸ‡¸` },
    'us-south1': { location: `Dallas, Texas, North America`, flag: `ğŸ‡ºğŸ‡¸` },
    'us-west1': { location: `The Dalles, Oregon, North America`, flag: `ğŸ‡ºğŸ‡¸` },
    'us-west2': {
        location: `Los Angeles, California, North America`,
        flag: `ğŸ‡ºğŸ‡¸`,
    },
    'us-west3': { location: `Salt Lake City, Utah, North America`, flag: `ğŸ‡ºğŸ‡¸` },
    'us-west4': { location: `Las Vegas, Nevada, North America`, flag: `ğŸ‡ºğŸ‡¸` },
};

export const getProviderLocationDetails = (
    provider: string | undefined,
    region: string | undefined
) => {
    if (!provider || !region) {
        return null;
    }

    if (provider.toUpperCase() === 'AWS') {
        return aws[region] ?? defaultResponse;
    }

    if (provider.toUpperCase() === 'GCP') {
        return gcp[region] ?? defaultResponse;
    }

    if (provider.toUpperCase() === 'AZURE' || provider.toUpperCase() === 'AZ') {
        return azure[region] ?? defaultResponse;
    }

    return defaultResponse;
};
