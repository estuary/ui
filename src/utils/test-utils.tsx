/* eslint-disable @typescript-eslint/unbound-method */
// https://testing-library.com/docs/react-testing-library/setup#custom-render

import { render as rtlRender, RenderOptions } from '@testing-library/react';
import { AuthContext } from 'context/Auth';
import AppContent from 'context/Content';
import AppRouter from 'context/Router';
import AppTheme from 'context/Theme';
import { ReactElement } from 'react';
import { auth, AuthDetails } from 'services/auth';

// TODO - this does not really do anything because we hardcode
//   the username down below on the MockAuthProvider. However,
//   this will be needed in the future so leaving the work so far
//   so we can leverage this when wanting to test the "on load auth"
//   functionality
const loginAsUser = (username: string) => {
    const mockAuthDetails: AuthDetails = {
        session: {
            account_id: '',
            expires_at: '',
            token: '',
        },
        user: {
            created_at: '',
            display_name: username,
            email: `${username}@${username}`,
            id: '',
            name: username,
            unique_name: username,
            updated_at: '',
        },
    };

    auth.saveSession(mockAuthDetails.session);
    auth.saveUser(mockAuthDetails.user);
};

const logoutUser = () => {
    auth.removeAuthDetails();
};

const goTo = (route?: string, name?: string) => {
    window.history.pushState(
        {},
        name ? name : 'Test page',
        route ? route : '/'
    );
};

const waitForLoading = async () => {
    return Promise.resolve();
};

// Mocking AuthContext out basically skips all the loading screen and boostrapping
//  stuff that happens on load. This should be fine for unit testing as it
//  can help speed up tests a bit and makes things easier to deal with. This
//  does mean that we need to eventually write a basic test that handles the
//  account bootstrapping.
interface MockAuthProps {
    children: JSX.Element;
    username?: string;
}
const MockAuthProvider = ({ children, username }: MockAuthProps) => {
    const value = {
        login: jest.fn(() => {
            if (username) {
                loginAsUser(username);
            }
            return Promise.resolve();
        }),
        logout: jest.fn(() => {
            logoutUser();
            return Promise.resolve();
        }),
        user: username,
    };

    return (
        <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
    );
};

const customRender = async (
    ui: ReactElement,
    options: Omit<RenderOptions, 'wrapper'> & {
        route?: string;
        user?: string;
    }
) => {
    const { route, user } = options;

    goTo(route, 'Test Page');

    const view = rtlRender(
        <AppContent>
            <MockAuthProvider username={user}>
                <AppTheme>
                    <AppRouter>{ui}</AppRouter>
                </AppTheme>
            </MockAuthProvider>
        </AppContent>,
        {
            ...options,
        }
    );

    await waitForLoading();

    return view;
};

export * from '@testing-library/react';
export { goTo, loginAsUser, logoutUser, customRender, waitForLoading };
