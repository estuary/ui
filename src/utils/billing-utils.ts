import { isEqual, parseISO } from 'date-fns';
import { isEmpty, sum } from 'lodash';
import prettyBytes from 'pretty-bytes';
import { BillingRecord } from 'stores/Billing/types';
import { CatalogStats_Billing, Entity } from 'types';

export const TOTAL_CARD_HEIGHT = 300;

// Grid item height - 72 = graph canvas height
export const CARD_AREA_HEIGHT = TOTAL_CARD_HEIGHT - 72;

export const BYTES_PER_GB = 1073741824;

const FREE_TASK_COUNT = 2;

export enum FREE_GB_BY_TIER {
    FREE = 20,
    PERSONAL = 10,
    ENTERPRISE = 10,
}

export const evaluateSpecType = (query: CatalogStats_Billing): Entity => {
    if (Object.hasOwn(query.flow_document.taskStats, 'capture')) {
        return 'capture';
    } else if (Object.hasOwn(query.flow_document.taskStats, 'materialize')) {
        return 'materialization';
    } else {
        return 'collection';
    }
};

const evaluateTotalCost = (dataVolume: number, taskCount: number) => {
    const freeBytes = 10 * BYTES_PER_GB;

    const dataVolumeOverLimit =
        dataVolume > freeBytes ? dataVolume - freeBytes : 0;

    const taskCountOverLimit =
        taskCount > FREE_TASK_COUNT ? taskCount - FREE_TASK_COUNT : 0;

    return (
        (dataVolumeOverLimit / BYTES_PER_GB) * 0.75 + taskCountOverLimit * 20
    );
};

const evaluateDataVolume = (stats: CatalogStats_Billing[]): number => {
    const taskBytes: number[] = stats.map((query) => {
        if (Object.hasOwn(query.flow_document.taskStats, 'capture')) {
            return query.bytes_written_by_me;
        } else if (
            Object.hasOwn(query.flow_document.taskStats, 'materialize')
        ) {
            return query.bytes_read_by_me;
        } else {
            return query.bytes_written_by_me + query.bytes_read_by_me;
        }
    });

    return sum(taskBytes);
};

export const stripTimeFromDate = (date: string) => {
    const [truncatedDateStr] = date.split('T', 1);

    return parseISO(truncatedDateStr);
};

const getInitialBillingRecord = (date: string): BillingRecord => {
    const truncatedDate = stripTimeFromDate(date);

    return {
        date: truncatedDate,
        dataVolume: 0,
        taskCount: 0,
        totalCost: 0,
        pricingTier: null,
        taskRate: null,
        gbFree: null,
        includedTasks: null,
    };
};

// TODO (billing): Remove this helper function to translate data returned from
//   the new RPC when available.
export const formatBillingCatalogStats = (
    value: CatalogStats_Billing[]
): BillingRecord[] => {
    const taskStatData = value.filter((query) =>
        Object.hasOwn(query.flow_document, 'taskStats')
    );

    let sortedStats: {
        [date: string]: CatalogStats_Billing[];
    } = {};

    taskStatData.forEach((query) => {
        if (Object.hasOwn(sortedStats, query.ts)) {
            sortedStats[query.ts].push(query);
        } else {
            sortedStats = {
                ...sortedStats,
                [query.ts]: [query],
            };
        }
    });

    const billingHistory: BillingRecord[] = [];

    if (!isEmpty(sortedStats)) {
        Object.entries(sortedStats).forEach(([ts, stats]) => {
            const billingRecordIndex = billingHistory.findIndex((record) =>
                isEqual(record.date, stripTimeFromDate(ts))
            );

            const taskCount = stats.length;
            const dataVolume = evaluateDataVolume(stats);
            const totalCost = evaluateTotalCost(dataVolume, taskCount);

            if (billingRecordIndex === -1) {
                const { date, pricingTier, taskRate, gbFree, includedTasks } =
                    getInitialBillingRecord(ts);

                billingHistory.push({
                    date,
                    dataVolume,
                    taskCount,
                    totalCost,
                    pricingTier: pricingTier ?? 'personal',
                    taskRate: taskRate ?? 20,
                    gbFree: gbFree ?? FREE_GB_BY_TIER.PERSONAL,
                    includedTasks: includedTasks ?? 2,
                });
            } else {
                const { date, pricingTier, taskRate, gbFree, includedTasks } =
                    billingHistory[billingRecordIndex];

                billingHistory[billingRecordIndex] = {
                    date,
                    dataVolume,
                    taskCount,
                    totalCost,
                    pricingTier: pricingTier ?? 'personal',
                    taskRate: taskRate ?? 20,
                    gbFree: gbFree ?? FREE_GB_BY_TIER.PERSONAL,
                    includedTasks: includedTasks ?? 2,
                };
            }
        });
    }

    return billingHistory;
};

export interface SeriesConfig {
    data: [string, number][];
    seriesName?: string;
    stack?: string;
}

export const formatDataVolumeForDisplay = (
    seriesConfigs: SeriesConfig[],
    tooltipConfig: any
): string => {
    const filteredSeries =
        seriesConfigs.length === 1
            ? seriesConfigs
            : seriesConfigs.filter(
                  (series) => series.seriesName === tooltipConfig.seriesName
              );

    const dataVolumeInBytes = filteredSeries
        .flatMap(({ data }) => data)
        .find(([month]) => month === tooltipConfig.name);

    return dataVolumeInBytes
        ? prettyBytes(dataVolumeInBytes[1], { minimumFractionDigits: 2 })
        : `${tooltipConfig.value[1]} GB`;
};
