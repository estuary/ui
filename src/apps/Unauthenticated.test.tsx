import '@testing-library/jest-dom';
import { customRender } from '../utils/test-utils';
import Unauthenticated from './Unauthenticated';

test('Unauthenticated app renders correctly', async () => {
    const { container } = await customRender(<Unauthenticated />, {
        route: '/login',
    });

    expect(container).toMatchSnapshot();
});
