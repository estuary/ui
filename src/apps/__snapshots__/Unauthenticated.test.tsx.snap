// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unauthenticated app renders correctly 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-zyhqkc-MuiGrid-root"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1mifq5q-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-dv03od-MuiToolbar-root"
      >
        <div
          class="MuiBox-root css-zdpt2t"
        >
          <div
            class="MuiBox-root css-ji4awn"
          >
            <img
              alt="Estuary"
              src="logo-estuary.png"
              width="120"
            />
          </div>
        </div>
        <div
          class="css-1f39f1a-MuiStack-root"
        >
          <button
            aria-controls="help-menu-menu"
            aria-haspopup="true"
            aria-label="Open help"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-j8rmhe-MuiButtonBase-root-MuiIconButton-root"
            data-mui-internal-clone-element="true"
            id="help-menu-button"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="HelpIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 17h-2v-2h2v2zm2.07-7.75-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25z"
              />
            </svg>
          </button>
        </div>
      </div>
    </header>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-1dz4rvg-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiCard-root css-brevj6-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-1nsgtsg-MuiTypography-root"
          >
            Control Plane
          </h2>
        </div>
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          When running locally you can login with whatever name you want.
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-10r5to9-MuiCardActions-root"
        >
          <form>
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-required css-1jetl7u-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="userName"
                id="userName-label"
              >
                User Name
                <span
                  aria-hidden="true"
                  class="MuiInputLabel-asterisk MuiFormLabel-asterisk css-1h537sd-MuiFormLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-9op6k-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="userName"
                  required=""
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-173wfxe"
                  >
                    <span>
                      User Name
                       
                      *
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-u53hp6-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="submit"
            >
              Login
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
